monty-31/src/aarch64_neon/poseidon2.rs

```diff
@@ -318,13 +350,78 @@ where
     fn permute_state_terminal(&self, state: &mut [PackedMontyField31Neon<FP>; WIDTH]) {
         external_terminal_permute_state(
             state,
-            self.external_constants.get_terminal_constants(),
-            add_rc_and_sbox_generic,
+            &self.packed_terminal_external_constants,
+            add_rc_and_sbox::<FP, D>,
             &MDSMat4,
         );
     }
 }
 
+/// Converts a scalar constant into a packed NEON vector in "negative form".
+///
+/// This is a pre-computation step for round constants. Instead of storing a constant `c`
+/// and performing a modular addition `(val + c) mod P` in the permutation loop, we pre-compute
+/// `c' = c - P`. This allows the round update to be a simple, non-modular signed integer
+/// addition `val + c'`, which is significantly faster.
```

Comment:
This mixes what the function does with why it does that. The latter is usually better as an inline comment in the appropriate place.
```suggestion
/// Instead of storing a constant `c`, we pre-compute `c' = c - P` and store it as a packed vector.
```